src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2565,39 +2565,30 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char
             // We need to chop the output string if it is too long.
             int cursor = 0; // This records the chopping position in output string
             const int MaxBufferSize = 16383; // this is 64K/4 - 1 to account for possible width of each character.
+            ReadOnlySpan<char> outBuffer;
 
-            while (cursor < output.Length)
+            while (cursor + MaxBufferSize < output.Length)
             {
-                ReadOnlySpan<char> outBuffer;
-
-                if (cursor + MaxBufferSize < output.Length)
-                {
-                    outBuffer = output.Slice(cursor, MaxBufferSize);
-                    cursor += MaxBufferSize;
+                outBuffer = output.Slice(cursor, MaxBufferSize);
+                cursor += MaxBufferSize;
+                WriteConsole(consoleHandle, outBuffer);
+            }
 
-                    WriteConsole(consoleHandle, outBuffer);
-                }
-                else
-                {
-                    outBuffer = output.Slice(cursor);
-                    cursor = output.Length;
+            outBuffer = output.Slice(cursor);
+            cursor = output.Length;
 
-                    if (newLine)
-                    {
-                        var endOfLine = Environment.NewLine.AsSpan();
-                        var endOfLineLength = endOfLine.Length;
-#pragma warning disable CA2014
-                        Span<char> outBufferLine = stackalloc char[outBuffer.Length + endOfLineLength];
-#pragma warning restore CA2014
-                        outBuffer.CopyTo(outBufferLine);
-                        endOfLine.CopyTo(outBufferLine.Slice(outBufferLine.Length - endOfLineLength));
-                        WriteConsole(consoleHandle, outBufferLine);
-                    }
-                    else
-                    {
-                        WriteConsole(consoleHandle, outBuffer);
-                    }
-                }
+            if (newLine)
+            {
+                var endOfLine = Environment.NewLine.AsSpan();
+                var endOfLineLength = endOfLine.Length;
+                Span<char> outBufferLine = stackalloc char[outBuffer.Length + endOfLineLength];
```

Comment:
Here we could allocate ~16KB (MaxBufferSize) in stack. That it is not good. I suggest to output the tail before the condition if the tail >= MaxStackAlloc ( MaxStackAlloc = 512).

Reply:
@iSazonov 
I think what you are suggesting is that we should further split the `outBuffer` into 2 pieces -> MaxBufferSize - 512b and 512b, such that the span that is stacalloc'ed is 512 bytes.
(assuming that the size of the `outBuffer` is `MaxBufferSize`, in the worst case)

The first MaxBufferSize - 512b piece will need additional memory. This will also end up going on the stack since we are using spans. This route does not seem any better than allocating the entire MaxBufferSize.

Also, this will mean that we would now need to make two `WriteConsole` calls to finish writing the contents of the outBuffer, rather than 1. 
I don't think the memory allocation issue is significant enough to justify adding the perf of a whole other IO operation.

Let me know if I understood that right.

Reply:
I mean:
```c#
        internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char> output, bool newLine)
        {
            Dbg.Assert(!consoleHandle.IsInvalid, "ConsoleHandle is not valid");
            Dbg.Assert(!consoleHandle.IsClosed, "ConsoleHandle is closed");

            const char MaxStackAlloc = 512;
            Span<char> buffer = stackalloc char[MaxStackAlloc];
            ReadOnlySpan<char> outBuffer;

            // Native WriteConsole doesn't support output buffer longer than 64K.
            // We need to chop the output string if it is too long.
            // This records the chopping position in output string
            int cursor = 0;

            // this is 64K/4 - 1 to account for possible width of each character.
            const int MaxBufferSize = 16383;

            while (cursor <= output.Length - MaxBufferSize)
            {
                outBuffer = output.Slice(cursor, MaxBufferSize);
                cursor += MaxBufferSize;

                WriteConsole(consoleHandle, outBuffer);
            }

            if (cursor == output.Length)
            {
                if (newLine)
                {
                    WriteConsole(consoleHandle, Environment.NewLine);
                }

                return;
            }

            outBuffer = output.Slice(cursor);

            if (!newLine)
            {
                WriteConsole(consoleHandle, outBuffer);

                return;
            }

            if (outBuffer.Length < MaxStackAlloc - 1)
            {
                // We expect it is a hot path.
                var endOfLine = Environment.NewLine.AsSpan();
                var endOfLineLength = endOfLine.Length;
                buffer = buffer.Slice(0, outBuffer.Length + endOfLineLength);            }

                outBuffer.CopyTo(buffer);
                endOfLine.CopyTo(buffer.Slice(buffer.Length - endOfLineLength));

                WriteConsole(consoleHandle, buffer);
            }
            else
            {
                WriteConsole(consoleHandle, outBuffer);

                WriteConsole(consoleHandle, Environment.NewLine);
            }
        }
```

Reply:
Up.

Reply:
@iSazonov I understand your concern here. But the `while (cursor + MaxBufferSize < output.Length)` comes from the existing `if (cursor + MaxBufferSize < output.Length)`, so it's not a problem introduced by this cleanup change. So, I'm fine merging this PR as is, and submit a new PR to address your concern.

Reply:
@daxian-dbw Agree.